### ISSUE TITLE
What is the intuition behind setting different seed between epochs?

### ISSUE THREAD

**Issue by VinhLoiIT** [pytorch/ignite#851]
## ❓ Questions

When running the engine, it produces different results each time. I have set the seed globally:
```python
seed = 0
torch.manual_seed(seed)
torch.cuda.manual_seed_all(seed)
import numpy as np
np.random.seed(seed)
import random
random.seed(seed)
```
However, the result is slightly different:
This is the results between two runs
![image](https://user-images.githubusercontent.com/8319105/77654937-4a4b7400-6fa4-11ea-8b97-dd0660432b8b.png)
And this is when I zoom in:
![image](https://user-images.githubusercontent.com/8319105/77655777-71567580-6fa5-11ea-8d46-747ba2315e6b.png)
The code does not change anything, just run twice.

I see that Engine have change the seed every epoch right? what is the intuition behind it? Does not a single global seed is enough? If not, how to reproduce the result?

**Comment by vfdev-5** [comment]
> I see that Engine have change the seed every epoch right? what is the intuition behind it? Does not a single global seed is enough? If not, how to reproduce the result?

@VinhLoiIT idea is to synchronize randomness of dataflow at each epoch. This helps to tackle the problem to resume checkpointed training : https://pytorch.org/ignite/concepts.html#resume-training

However, today this has some issues (eg #795), in 0.4.0 we probably let user to disable this and make it optional.


> However, the result is slightly different

Are you using cuda with benchmark = True ? Please, also take a look at : https://pytorch.org/docs/stable/notes/randomness.html#cudnn

**Comment by VinhLoiIT** [comment]
> Are you using cuda with benchmark = True ? Please, also take a look at : https://pytorch.org/docs/stable/notes/randomness.html#cudnn

Thank you, this solved the problem. After adding:
```python
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False
```
at the beginning of the program, it runs exactly the same.

Thank you
